File: 659-split-array-into-consecutive-subsequences/659-split-array-into-consecutive-subsequences.py
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        to_be_placed = collections.Counter(nums);
        endingAt = collections.Counter();
        
        for num in nums:
            if not to_be_placed[num]:
                continue;
            if endingAt[num-1]>0:
                endingAt[num-1]-=1;
                endingAt[num]+=1;
            elif to_be_placed[num+1]>0 and to_be_placed[num+2]>0:
                to_be_placed[num+1]-=1;
                to_be_placed[num+2]-=1;
                endingAt[num+2]+=1;
            else:
                return False;
            to_be_placed[num]-=1;
        return True;

File: 659-split-array-into-consecutive-subsequences/659-split-array-into-consecutive-subsequences.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums) {
        int n = nums.size();
        map<int,priority_queue<int,vector<int>,greater<int>>>mpp;
        for(int i=0;i<n;i++){
            if(mpp.find(nums[i]-1)!=mpp.end()){
                int a = mpp[nums[i]-1].top();
                mpp[nums[i]-1].pop();
                if(mpp[nums[i]-1].size()==0){
                    mpp.erase(nums[i]-1);
                }
                mpp[nums[i]].push(a+1);
            }else{
                mpp[nums[i]].push(1);
            }
        }
        for(auto it:mpp){
            if(it.second.size() and it.second.top()<3) return false;
        }
        return true;
    }
};